File: sol/1462/1462.cpp
#include <queue>
#include <unordered_map>
#include <unordered_set>

struct Solution {
    vector<bool> checkIfPrerequisite(int numCourses, vector<vector<int>>& prerequisites, vector<vector<int>>& queries) {
        unordered_map<int, vector<int>> edges(numCourses);
        unordered_map<int, int> indeg(numCourses);
        for (auto &e : prerequisites){
            edges[e[0]].push_back(e[1]);
            ++indeg[e[1]];
        }
        queue<int> buff;
        for (int i = 0; i < numCourses; ++i){
            if (indeg[i] == 0){
                buff.push(i);
            }
        }
        vector<unordered_set<int>> req(numCourses);
        while (!buff.empty()){
            int x = buff.front();
            buff.pop();
            for (int nb : edges[x]){
                req[nb].insert(req[x].begin(), req[x].end());
                req[nb].insert(x);
                if (--indeg[nb] == 0){
                    buff.push(nb);
                }
            }
        }
        vector<bool> res(queries.size());
        for (int i = 0; i < queries.size(); ++i){
            res[i] = req[queries[i][1]].contains(queries[i][0]);
        }
        return res;
    }
};

File: sol/1462/1462.py
from collections import defaultdict, deque

class Solution:
    def checkIfPrerequisite(self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:
        edges = defaultdict(list)
        indeg = defaultdict(int)
        for a, b in prerequisites:
            edges[a].append(b)
            indeg[b] += 1
        buff = deque([a for a in range(numCourses) if indeg[a] == 0])
        req = [set() for _ in range(numCourses)]
        while len(buff) != 0:
            x = buff.popleft()
            for nb in edges[x]:
                req[nb].update(req[x])
                req[nb].add(x)
                indeg[nb] -= 1
                if indeg[nb] == 0:
                    buff.append(nb)
        return [a in req[b] for a, b in queries]
